blog/2025-09-26-sglang-ant-group.md

```diff
@@ -0,0 +1,292 @@
+
+### Deployment Strategy
+
+<img src="/images/blog/ant-group-prac/deploy.svg" style="display:block; margin-top: auto; margin-left: auto; margin-right: auto; margin-bottom: auto; width: 90%; image-orientation: none;"></img>
+
+
+#### Prefill
+- **SLA:** Prefill is compute-intensive, and multi-node DP+EP can inflate time-to-first-token (TTFT), often violating SLAs. A single-node TP setup keeps TTFT within target.
+- **Elastic Scaling:** Prefill must scale in and out with the KV cache. Single-node TP makes scaling straightforward, while multi-node DP+EP complicates resource and cache management.
+
+#### Decode
+- **Hardware Characteristics:** H20 trades compute for larger memory and higher NVLink bandwidth(compared with H800), enabling efficient KV-cache use and keeping MoE communication on high-bandwidth NVLink. 
+- **Fault Radius:** Smaller EP configurations limit the impact of decoding or GPU failures. With EP high-availability (HA) still maturing, smaller EP is safer and more reliable in production.
+
+### Optimizations
+
+#### Prefill
+
+<img src="/images/blog/ant-group-prac/prefill_overview.svg" style="display:block; margin-top: auto; margin-left: auto; margin-right: auto; margin-bottom: auto; width: 90%; image-orientation: none;"></img>
+
+##### Observation
+- MLA is costlier than MHA for long sequences.
+- MoE latency was unexpectedly high despite lower computation
+- `embed/mlp all reduce + RMSNorm + fused_qkv_a_proj_with_mqa` introduces redundant communication and computation in TP
+
+##### Solution
+- [MHA/MLA](https://github.com/sgl-project/sglang/pull/9551): Introduced tunable parameter `se = extend × (extend + prefix)` to select MHA or MLA based on batch size and sequence lengths.
+- [MoE](https://github.com/sgl-project/sglang/pull/10567): Optimized `b_scale` calculation, refactored input access of `down proj` with TMA, and tuned configurations based on real expert distributions.
+- [TP Optimization](https://github.com/sgl-project/sglang/pull/10568): Optimized `embed/mlp reduce scatter + RMSNorm + fused_qkv_a_proj_with_mqa + all gather` to reduce computation and communication.
+
+#### Decode
+##### Load Balance
+###### [Expert Affinity EPLB](https://github.com/antgroup-infra/sglang/pull/2)
+
+<img src="/images/blog/ant-group-prac/eplb.svg" style="display:block; margin-top: auto; margin-left: auto; margin-right: auto; margin-bottom: auto; width: 90%; image-orientation: none;"></img>
+
+Standard EPLB balances intra-GPU loads but overlooks correlations between experts, which often scatters frequently co-activated experts across nodes and increases cross-node communication overhead.    
+
+We extend EPLB by tracking **top-k expert co-activations** to build an **expert affinity matrix**. 
+After intra-GPU load balancing, we adjust placement so that **highly co-activated experts** are kept within the same node, thereby reducing cross-node communication and delivering an additional **~5% performance gain** over vanilla EPLB.  
```

Comment:
```suggestion
After intra-GPU load balancing, we adjust placement so that **highly co-activated experts** are kept within the same node, thereby reducing cross-node communication, delivering an additional **~5% performance gain** over vanilla EPLB.  
```